amethyst_assets/src/reload.rs

```diff
@@ -1,18 +1,115 @@
-//! MODULE DISABLED, will be reused later!
+//! Defines the `Reload` trait.
 
-use Source;
+use std::sync::Arc;
 
-pub trait Reload {
-    fn needs_reload(&self, source: &Source) -> bool;
+use {Asset, BoxedErr, Format, Source};
+
+/// The `Reload` trait provides a method which checks if an asset needs to be reloaded.
+pub trait Reload<A: Asset>: ReloadClone<A> + Send + Sync + 'static {
+    /// Checks if a reload is necessary.
+    fn needs_reload(&self) -> bool;
+    /// Returns the asset name.
+    fn name(&self) -> String;
+    /// Returns the format name.
+    fn format(&self) -> String;
+    /// Reloads the asset.
+    fn reload(self: Box<Self>) -> Result<(A::Data, Option<Box<Reload<A>>>), BoxedErr>;
+}
+
+pub trait ReloadClone<A> {
+    fn cloned(&self) -> Box<Reload<A>>;
+}
+
+impl<A, T> ReloadClone<A> for T
+where
+    A: Asset,
+    T: Clone + Reload<A>,
+{
+    fn cloned(&self) -> Box<Reload<A>> {
+        Box::new(self.clone())
+    }
 }
 
-pub struct SingleFile {
+impl<A: Asset> Clone for Box<Reload<A>> {
+    fn clone(&self) -> Self {
+        self.cloned()
+    }
+}
+
+/// An implementation of `Reload` which just stores the modification time
+/// and the path of the file.
+pub struct SingleFile<A: Asset, F: Format<A>> {
+    format: F,
     modified: u64,
+    options: F::Options,
     path: String,
+    source: Arc<Source>,
+}
+
+impl<A: Asset, F: Format<A>> SingleFile<A, F> {
+    /// Creates a new `SingleFile` reload object.
+    pub fn new(
+        format: F,
+        modified: u64,
+        options: F::Options,
+        path: String,
+        source: Arc<Source>,
+    ) -> Self {
+        SingleFile {
+            format,
+            modified,
+            options,
+            path,
+            source,
+        }
+    }
+}
+
+impl<A, F> Clone for SingleFile<A, F>
+where
+    A: Asset,
+    F: Clone + Format<A>,
+    F::Options: Clone,
+{
+    fn clone(&self) -> Self {
+        SingleFile {
+            format: self.format.clone(),
+            modified: self.modified,
+            options: self.options.clone(),
+            path: self.path.clone(),
+            source: self.source.clone(),
+        }
+    }
 }
 
-impl Reload for SingleFile {
-    fn needs_reload(&self, source: &Source) -> bool {
-        self.modified != 0 && source.modified(&self.path).unwrap_or(0) > self.modified
+impl<A, F> Reload<A> for SingleFile<A, F>
+where
+    A: Asset,
+    F: Clone + Format<A> + Sync,
+    <F as Format<A>>::Options: Clone + Sync,
+{
+    fn needs_reload(&self) -> bool {
+        self.modified != 0 && (self.source.modified(&self.path).unwrap_or(0) > self.modified)
+    }
+
+    fn reload(self: Box<Self>) -> Result<(A::Data, Option<Box<Reload<A>>>), BoxedErr> {
+        let this: SingleFile<_, _> = *self;
+        let SingleFile {
+            format,
+            path,
+            source,
+            options,
+            ..
+        } = this;
+
+        format.import(path, source, options, true)
```

Comment:
Can't it ask not to create new `Reload` object and return `self`?

Reply:
No, imagine a GLTF file which points to other files; in case the GLTF file gets reloaded, new files might have been added which we also want to track.

Reply:
But this is `SingleFile`

Reply:
So how would you do that?